### PR TITLE
Import Rust SDK as Experimental

### PR DESCRIPTION
**Pull request by gerhard**
This imports https://github.com/kjuulh/dagger-sdk under `sdk/rust` as an **experimental** SDK.

## @gerhard action items (part of this PR)
- [x] Open PR with https://github.com/kjuulh/dagger-sdk/ history import
	- [x] Capture commands used for history import
- [x] Give Kasper write permissions to `dagger` repository
- [x] Add @kjuulh to [`CODEOWNERS`](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) (`@dagger/sdk-rust` group)
  - If we like this, I propose that we do the same for other SDKs cc @shykes @samalba @aluzzardi 
- [x] Mention in SDK's README that Rust is currently experimental

## @kjuulh action items (part of this PR)
- [x] Configure `Sign-off-by` https://docs.dagger.io/204441/contributing/#dco
  - @gerhard manually set this to **Pass** so that @kjuulh does not need to rewrite `255` commits
- [x] Move (and rename) GHA workflows from `sdk/rust/.github` to `.github/`
- [x] Ensure all GHA workflows pass
- [ ] **Maybe** rewrite commits to fix link to issues / PRs
	- [ ] We should use absolutes, e.g. https://github.com/kjuulh/dagger-sdk/pull/61 instead of `#61`

## Follow-ups (outside / after this PR)
This sounds like maintainer on-boarding to me @kjuulh 🙂

- Migrate [open issues](https://github.com/kjuulh/dagger-sdk/issues) which are still relevant - `7` open currently
	- ~~Ensure that we have a new `sdk/rust` label~~ - thank you @d3rp3tt3!
	- cc @mircubed 
- Migrate PRs which are still relevant
- Add commit to https://github.com/kjuulh/dagger-sdk/ that removes all files except the README
	- Add repo migration notice to `README.md`
	- Archive repository in GitHub, e.g. https://github.com/thechangelog/infrastructure
- Tag `sdk/rust/v*` to check that a new release gets created as expected
  - This will require us to check that all three crates are published `by Dagger`. FTR: https://lib.rs/crates/dagger-sdk, https://lib.rs/crates/dagger-core & https://lib.rs/crates/dagger-codegen
- Add Rust SDK to https://github.com/dagger/dagger/blob/main/RELEASING.md
  - cc @gerhard 
- Add Rust SDK to https://docs.dagger.io/
	- Add recipe variant to https://docs.dagger.io/7442989/cookbook
	- Add guide variant to https://docs.dagger.io/278912/guides
	- cc @vikram-dagger 
- Integrate with Analytics
  - cc @grouville @d3rp3tt3 

Did I miss anything @shykes @samalba @kjuulh @d3rp3tt3 @mircubed?

---
BTW, I noticed that not all your commits @kjuulh are **Verified**:
Screenshot 2023-05-17 at 16 06 37

---
Fixes DAG-1028

### CODE REVIEW

**Comment by kjuulh** [comment]
Awesome. No idea why they arent signed probably something to do with github rebases.

I will give it a shot fixing the ci so that it is compatible with the rest :)

**Comment by gerhard** [comment]
Giving you write permissions next. Afterwards, you should be able to push to my fork, we do that all the time when we collaborate.

**Comment by gerhard** [comment]
Once you accept your invite @kjuulh, you should have write access to this repo:

image

Being familiar with this is important: https://docs.dagger.io/204441/contributing - especially the part about **always** committing to your `dagger` fork & submitting PRs from your fork. FWIW, I also opened this PR from my `dagger` fork.

**Comment by gerhard** [comment]
I have imported all git history from https://github.com/kjuulh/dagger-sdk using the following commands:

```console
git remote add dagger-sdk-rust git@github.com:kjuulh/dagger-sdk.git
git fetch dagger-sdk-rust --no-tags
git checkout -b sdk/rust dagger-sdk-rust/main
mkdir -p sdk/rust
git ls-tree -z --name-only HEAD | xargs -0 -I {} git mv {} sdk/rust/
git commit --gpg-sign --signoff

git checkout main
git pull
git checkout -b import-sdk-rust

git merge --allow-unrelated-histories sdk/rust

git push -u fork
```

This was helpful: https://stackoverflow.com/a/20974621

---
FWIW:

```console
git remote -v | rg "kjuulh|fork|origin"

dagger-sdk-rust	git@github.com:kjuulh/dagger-sdk.git (fetch)
dagger-sdk-rust	git@github.com:kjuulh/dagger-sdk.git (push)
fork	git@github.com:gerhard/dagger.git (fetch)
fork	git@github.com:gerhard/dagger.git (push)
origin	git@github.com:dagger/dagger.git (fetch)
origin	git@github.com:dagger/dagger.git (push)
```

**Comment by gerhard** [comment]
Anything else that we may want to add about the experimental nature of this SDK @shykes?

image

File: sdk/rust/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+> **Warning**
+> This is an **experimental** SDK.
```
Comment (shykes) [inline sdk/rust/README.md:2]:
```suggestion
> This SDK is **experimental**. Please do not use it for anything mission-critical. Possible issues include:

* Missing features
* Stability issues
* Performance issues
* Lack of polish
* Upcoming breaking changes
* Incomplete or out-of-date documentation

Please report any issues you encounter. We appreciate and encourage contributions. If you are a Rust developer interested in contributing to this SDK, we welcome you!
```

File: sdk/rust/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Architecture
```
Comment (shykes) [inline sdk/rust/ARCHITECTURE.md:1]:
```suggestion
**Warning**
This SDK is **experimental**. Please do not use it for anything mission-critical. Possible issues include:
* Missing features
* Stability issues
* Performance issues
* Lack of polish
* Upcoming breaking changes
* Incomplete or out-of-date documentation
Please report any issues you encounter. We appreciate and encourage contributions. If you are a Rust developer interested in contributing to this SDK, we welcome you!

# Architecture
```

**Comment by kjuulh** [comment]
I am pretty much done with the upstreaming. I only really need to fix the provisioning step and adding the experimental cli support. I expect there to be a few hours of work left and then I would consider this done.

I will probably also go back and add the signed-off-by, so that we can avoid the DCO excemption.

**Comment by kjuulh** [comment]
that said I gotta fix the test job for rust. the last run took 19min; which is absolutely wild

**Comment by gerhard** [comment]
That's awesome @kjuulh!

Notice that we use larger GitHub Actions runners for jobs which benefit from more CPUs & memory. The plan is to add "magicache" and/or managed Dagger Engines so that jobs wouldn't always start with a cold Dagger Engine.

**Review by gerhard — COMMENTED** [review]
Did a pass over all files. Left comments online. This feels close now.

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -154,6 +155,14 @@ jobs:
...
           yarn test -g 'Automatic Provisioned CLI Binary'
 
+      - name: "Setup Rust"
+        uses: dtolnay/rust-toolchain@stable
+
+      - name: "Test Rust SDK"
+        run: |
+          cd sdk/rust
+          cargo test --all
+
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply .github/workflows/publish.yml:165]: I forgot to comment it out: but it follows the same reasoning as this comment:

https://github.com/dagger/dagger/pull/5148/files/2ba94ec9d34c0f1e69eeec6af681c9255c0e0bd6#diff-551d1fcf87f78cc3bc18a7b332a4dc5d8773a512062df881c5aba28a6f5c48d7R314-R315

File: sdk/rust/.github/workflows/ci-multiplatform.yml [hunk 4]
```diff
@@ -0,0 +1,150 @@
+name: ci-multi-platform
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply sdk/rust/.github/workflows/ci-multiplatform.yml:1]: forgot to remove this. This is already handled by rust:sdk:test

File: sdk/rust/.github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+name: ci
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply sdk/rust/.github/workflows/ci.yaml:1]: forgot to remove this. This is already handled by rust:sdk:test

**Comment by kjuulh** [comment]
I will follow up and remove the .github things from `sdk/rust`, and remove all mentions of the old repo. there isn't a huge need to support other architectures just yet, and is something that can be handled as we want in the hack/sdk/tests instead.

File: sdk/rust/.github/workflows/lint-pr.yml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+      - edited
+      - synchronize
+jobs:
+  main:
+    name: Validate PR title
+    runs-on: ubuntu-latest
+    steps:
+      - uses: amannn/action-semantic-pull-request@v5
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply sdk/rust/.github/workflows/lint-pr.yml:15]: forgot to remove this. This isn't needed as we follow daggers release process (i.e. we don't follow strict semver)

[the remark this answers is not in the document]
Comment (gerhard) [reply sdk/rust/.github/workflows/lint-pr.yml:15]: I missed the `sdk/rust` part. All good!

File: sdk/rust/.github/workflows/ci.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (gerhard) [reply sdk/rust/.github/workflows/ci.yaml]: I missed the `sdk/rust` path. All good! 👍

**Comment by kjuulh** [comment]
@gerhard I would pretty much call this done. I did a bit of further cleanup, but the rest can wait. I had to set a bigger machine for lint and build. (A rust lint check requires a build...), feel free to revert this if needed, I will fix it in a follow-up pr. where I add the experimental support for cli.

**Comment by gerhard** [comment]
Thank you @kjuulh for seeing this through! The follow-up PRs should be a lot quicker to get through now that we have the big one out of the way.

FTR I went for an exceptional merge commit since we want to preserve the original history of https://github.com/kjuulh/dagger-sdk. cc @aluzzardi @samalba @shykes

**Comment by mircubed** [comment]
Note - I discussed this with @vikram-dagger and the documentation will be a heavy lift. We should keep this in mind as we plan the GTM around this.

**Comment by gerhard** [comment]
I have just removed all mentions of docs from the PR description @shykes.

**Comment by gerhard** [comment]
Any chance of actioning the following @kjuulh?
* Add commit to https://github.com/kjuulh/dagger-sdk/ that removes all files except the README
  * Add repo migration notice to README.md
  * Archive repository in GitHub, e.g. https://github.com/thechangelog/infrastructure

Here is what the Elixir SDK did: https://github.com/wingyplus/dagger_ex

**Comment by kjuulh** [comment]
@gerhard sure. I completely forgot =D

**Comment by kjuulh** [comment]
It is done =D

**Pull request merged**